src/Giraffe/WebSockets.fs

```diff
@@ -0,0 +1,182 @@
+                    do! this.WebSocket.SendAsync(segment, WebSocketMessageType.Text, true, cancellationToken)
+        }
+        
+
+        /// Closes the connection to the WebSocket client.
+        member this.CloseAsync(?reason,?cancellationToken) = task {
+            if not (isNull this.WebSocket) then
+                if this.WebSocket.State = WebSocketState.Open then
+                    let cancellationToken = cancellationToken |> Option.defaultValue CancellationToken.None
+                    let reason = reason |> Option.defaultValue "Closed by the WebSocket server"
+                    do! this.WebSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, reason, cancellationToken)
+        }
+
+        
+        /// Creates a new reference to a WebSocket.
+        static member FromWebSocket(websocket,?webSocketID,?subProtocol) : WebSocketReference = {
+            WebSocket  = websocket
+            Subprotocol = subProtocol
+            ID = webSocketID |> Option.defaultWith (fun _ -> Guid.NewGuid().ToString())
+        }
+
+/// A connection manager keeps track of all connections that are open at a specific endpoint.
+type ConnectionManager(?messageSize) =
+    let messageSize = defaultArg messageSize DefaultWebSocketOptions.ReceiveBufferSize
+
+    let connections = new System.Collections.Concurrent.ConcurrentDictionary<string, WebSocketReference>()
+
+    with
+
+        /// Tries to find a WebSocket by ID. Returns None if the socket wasn't found.
+        member __.TryGetWebSocket(websocketID:string) : WebSocketReference option = 
+            match connections.TryGetValue websocketID with
+            | true, r -> Some r
+            | _ -> None
+
+        /// Returns the number of WebSocket connections.
+        member __.Count = connections.Count
+
+        /// Sends a UTF-8 encoded text message to all WebSocket connections.
+        member __.BroadcastTextAsync(msg:string,?cancellationToken:CancellationToken) = task {
+            let byteResponse = System.Text.Encoding.UTF8.GetBytes msg
+            let segment = ArraySegment<byte>(byteResponse, 0, byteResponse.Length)
+            let cancellationToken = cancellationToken |> Option.defaultValue CancellationToken.None
+
+            let! _ =
+                connections
+                |> Seq.map (fun kv -> task {
+                    try
+                        if not cancellationToken.IsCancellationRequested then
+                            let webSocket = kv.Value.WebSocket
+                            if webSocket.State = WebSocketState.Open then
+                                do! webSocket.SendAsync(segment, WebSocketMessageType.Text, true, cancellationToken)
+                            else
+                                if webSocket.State = WebSocketState.Closed then
+                                    connections.TryRemove kv.Key |> ignore
+                    with
+                    | _ -> () })
+                |> Task.WhenAll
```

Comment:
This does not limit parallelism. Is it OK?

Reply:
The Sequence is generating tasks from connections enumeration to allow next socket to be called before the last task finished, as would be the case with for loop in task CE, Task.WhenAll ensures returned method task marks completed when all sockets have been completed (rather than fire&forget), where do you see the parallelism bottleneck?

Reply:
Since these should be I/O tasks I don't think there will be a problem.